### ISSUE TITLE
A problem about the clusters and pseudolabels every epoch

### ISSUE THREAD

**Issue by skr998** [facebookresearch/deepcluster#81]
After reading the code, I found that there seems to be no corresponding relationship about the cluster indexs(pseudolabels) between two epochs. Because the cluster index obtained by each epoch in the Kmeans is random,  how to associate the cluster indexs between two epochs? Otherwise, the calculation of cross-entropy as (y , pseudolabel) is meaningless. I am looking forward to your answer, thank you.

**Comment by enaserianhanzaei** [comment]
here is an explanation of how NMI is calculated https://scikit-learn.org/stable/modules/generated/sklearn.metrics.normalized_mutual_info_score.html

As you can see in the examples, and also explanation 'This metric is independent of the absolute values of the labels: a permutation of the class or cluster label values won’t change the score value in any way.'

**Comment by yyou1996** [comment]
The NMI (calculated during inference) is independent, but my confusion is that whether the reassignment without a corresponding relationship between epochs affects the training?

For example, I have 3 images (A, B, C) with 2 clusters (0, 1).
(1) In epoch 0, the cluster assignment is (0, 0, 1) for (A, B, C), so my CNN learn to assign images A, B with labels 0 and C with label 1.
(2) In epoch 1, the cluster assignment comes to (1, 1, 0). The NMI between epoch 0 & 1 assignment is high, but my CNN learns an opposite thing in epoch 1 that it tries to assign images A, B with labels 1 and C with label 0.

**Comment by skr998** [comment]
@yyou1996 This is my confusion, too. I think there are two possibility. First, they ignored this problem, after the program got a good performance at that time. Second, this problem about assignment has little impact on the result.

**Comment by yyou1996** [comment]
@enaserianhanzaei Would you mind providing some comments on this? It will greatly facilitate our deeper understanding of your excellent work. Thanks!

**Comment by enaserianhanzaei** [comment]
@yyou1996 It's not my work (I wish it was lol).

So, here is my understanding, NMI doesn't care about the label of the clusters, what it cares is whether the images belong to the same cluster in different epoches. 

Consider your own example:
"
For example, I have 3 images (A, B, C) with 2 clusters (0, 1).
(1) In epoch 0, the cluster assignment is (0, 0, 1) for (A, B, C), so my CNN learn to assign images A, B with labels 0 and C with label 1.
(2) In epoch 1, the cluster assignment comes to (1, 1, 0). The NMI between epoch 0 & 1 assignment is high, but my CNN learns an opposite thing in epoch 1 that it tries to assign images A, B with labels 1 and C with label 0.
"

you can see in both epoches, A and B belong to same cluster, and C in other cluster, the label of the clusters are different which doesn't matter. 

I hope it helps, and sorry for the late reply

**Comment by skr998** [comment]
@enaserianhanzaei Thank you for your reply. I want to ask that whether the images belong to the different labels in different epoches will have a negative effect on the results(CNN), because the network will learn this false representation.

**Comment by yyou1996** [comment]
@enaserianhanzaei Thanks a lot for the voluntary answer (I thought it was urs work lol). I think ur point makes sense but I cannot fully buy it. Hopefully I can hear others' opinions.

**Comment by mathildecaron31** [comment]
Hi @skr998 , @yyou1996 , @enaserianhanzaei 

@skr998  your point is totally valid, there is no mapping from the cluster assignments in two following epochs. That is why we need to reset the last layer at each new epoch.
This definitely is an issue and we propose to adresse this by replacing the classification into the pseudo-labels (given by cluster assignments) with direct comparison to k-means clustering in DeepCluster-v2 (see https://arxiv.org/pdf/2006.09882.pdf section and 5.1 as well as appendix D)

@enaserianhanzaei  I totally agree with your point `what it cares is whether the images belong to the same cluster in different epoches`.
